WordPress/gutenberg
# Autocomplete Component: add e2e tests (take two)

## What?
Add some basic e2e testing for the Autocomplete component, as well as some tests specific to past regressions

## Why?
The component currently lacks detailed testing, outside of some user mention tests. Increased testing will allow for safer/more confident updates to the component in the future.

## How?
This is an alternate approach based on the initial work and feedback in #42674

There are two specific regressions [introduced here](https://github.com/WordPress/gutenberg/pull/41382) (already reverted) that I want to include tests to avoid. See issues below for more details on those regressions:
- https://github.com/WordPress/gutenberg/issues/41709
- https://github.com/WordPress/gutenberg/issues/41724

These regression appear to effect only certain completers, like user mentions. I _think_ the issue is actually any autocompleter that relies on fetched data to populate its options. I haven't fully confirmed that theory, but the important part is that these regressions don't impact a completer whose options come from a simple locally declared object.

With that in mind I wanted these tests to confirm if a failure was impacting simple completers, more complex completers (like user mentions), or both. To accomplish this I'm running the tests through a `describe.each` to account for all cases. Because the data set (the completer options) is different for the two paths, most of the tests conditionally define their test data within the test itself. This way the testing logic is the same but the data is tailored to each completer. It should also make it relatively easy to test additional completers in the future if needed.

I also had to add a bit of logic for a single test in the `.each` to be skipped for user mentions. I wrapped this logic in a helper function so it can also be easily reused in the future if necessary.

Finally, in addition to creating a custom completer, the testing plugin also creates a clone of the existing user mention completer. This is used in one test to confirm that two complex (i.e. data-fetching) completers used in the same block won't cause the browser to crash (one of the regressions I'm trying to guard against). Unfortunately, while this cloning process works and accurately triggers the regression, the infinite loop at the heart of that regression _also_ crashes our tests 🤦. I haven't found a way around that, so I've left the test skipped for now. We can remove it (I know the linter disapproves of skipped tests) if that's preferable, but for now I opted to keep it in case we do find a way to utilize it while still getting test results in the event that this particular regression is ever reintroduced.

## Testing Instructions
1. Run `npm run test:e2e  -- packages/e2e-tests/specs/editor/various/autocomplete-and-mentions.test.js`
2. All tests should pass
3. Apply the changes from #41382 to re-introduce the regressions. Rebuild Gutenberg.
4. Re-run the tests. You should see one failure:
- User Mentions: should allow newlines after multiple completions

Optional: Finally, if you remove the `.skip` from the `Autocomplete: should insert elements from multiple completers in a single block` test and re-run the suite, you should find that it simply hangs, not returning any data. I'm calling this an optional step because we're really just testing that the test can trigger the regression by breaking your browser, which isn't actually all that helpful 😉
